providers/googledns/googledns.go

```diff
@@ -0,0 +1,197 @@
+package googledns
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/juju/ratelimit"
+	"github.com/rancher/external-dns/providers"
+	"github.com/rancher/external-dns/utils"
+	cnet "golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/dns/v1beta2"
+)
+
+var (
+	googlednsMaxRetries int    = 3
+	projectId           string = os.Getenv("GOOGLEDNS_PROJECT_ID")
+)
+
+type GoogleDNSProvider struct {
+	client       *dns.Service
+	hostedZoneId string
+	limiter      *ratelimit.Bucket
+}
+
+func init() {
+	providers.RegisterProvider("googledns", &GoogleDNSProvider{})
+}
+
+// Init creates a Google DNS client with credentials from DefaultClient, which uses FindDefaultCredentials
+
+// FindDefaultCredentials searches for "Application Default Credentials".
+//
+// It looks for credentials in the following places,
+// preferring the first location found:
+//
+//   1. A JSON file whose path is specified by the
+//      GOOGLE_APPLICATION_CREDENTIALS environment variable.
+//   2. A JSON file in a location known to the gcloud command-line tool.
+//      On Windows, this is %APPDATA%/gcloud/application_default_credentials.json.
+//      On other systems, $HOME/.config/gcloud/application_default_credentials.json.
+//   3. On Google App Engine it uses the appengine.AccessToken function.
+//   4. On Google Compute Engine and Google App Engine Managed VMs, it fetches
+//      credentials from the metadata server.
+//      (In this final case any provided scopes are ignored.)
+
+func (r *GoogleDNSProvider) Init(rootDomainName string) error {
+	r.limiter = ratelimit.NewBucketWithRate(5.0, 1)
+
+	if envVal := os.Getenv("GOOGLEDNS_MAX_RETRIES"); envVal != "" {
+		i, err := strconv.Atoi(envVal)
+		if err == nil {
+			googlednsMaxRetries = i
+		} else {
+			logrus.Warnf("Invalid value for GOOGLEDNS_MAX_RETRIES. Using default.")
+		}
+	}
+
+	ctx := cnet.Context(cnet.Background())
+
+	c, _ := google.DefaultClient(ctx, dns.CloudPlatformScope)
+
+	r.client, _ = dns.New(c)
+
+	if err := r.setHostedZone(rootDomainName); err != nil {
+		return fmt.Errorf("Failed to configure hosted zone: %v", err)
+	}
+
+	logrus.Infof("Configured %s with hosted zone %s",
+		r.GetName(), rootDomainName)
+
+	return nil
+}
+
+func (r *GoogleDNSProvider) setHostedZone(rootDomainName string) error {
+	r.limiter.Wait(1)
+
+	zones, err := r.client.ManagedZones.List(projectId).Do()
+
+	if len(zones.ManagedZones) == 0 {
+		return fmt.Errorf("Hosted zone for '%s' not found, got empty list", rootDomainName)
+	}
+
+	for _, zone := range zones.ManagedZones {
+		fmt.Printf("Found Zone: %v (%v)\n", zone.Name, zone.DnsName)
```

Comment:
I would rather not print this intermediate info. You can log it in Debug. Also I suggest you use logrus for logging